---
title: "SWNE trajectory inference 0829"
author: "Eric"
date: "2021/8/29"
output: 
  html_document:
    df_print: kable
    theme: united
  github_document:
    html_preview: true
    toc: true
    toc_depth: 3
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  tidy = T,
  echo = F,
  warning = F,
  message = F
)
knitr::opts_knit$set(root.dir = "i:/genomicdata/XSu-mouse/SWNE/0828")
```

```{r, echo=FALSE}
library(monocle3)
library(ggplot2)
library(ggrepel)
library(kableExtra)
Fibro <- readRDS("Fibro.rds")
Mye <- readRDS("Mye.rds")
Tumor <- readRDS("Tumor.rds")
```

## Some supplemental Information

> *Visualizing and Interpreting Single-Cell Gene Expression Datasets with Similarity Weighted Nonnegative Embedding*
> DOI: [10.1016/j.cels.2018.10.015](https://doi.org/10.1016/j.cels.2018.10.015)  

> *De Novo Prediction of Stem Cell Identity using Single-Cell Transcriptome Data*
> DOI: [10.1016/j.stem.2016.05.010](https://doi.org/10.1016/j.stem.2016.05.010)  

About the biological interpretation, SWNE article(the first) says, “One can also run Geneset Enrichment Analysis on the gene loadings for each factor to find the top genesets associated with that factor.” GSEA is calculated using fgseasimple (geneset permutation) and merged genesets. The geneset with lowest pvalue and highest NES is referred as the top geneset associated with that factor. Because some pathway names are too long, so we used the "F-number" to annotate the factor, and labeled the factor with corresponding pathway in the table. These pathway labels means they are mostly associated with this factor.  
	
  About the pseudotime calculation, it is necessary to provide a root where pseudotime equal to 0 no matter whether it is in monocle2 or 3. In monocle2, it will select a root "arbitrarily" if no root provided, while researchers must provide a root or a root area in monocle3. Here, enlightened by the second article, I calculated median entropy of each node using its nearest cell points, and chose node with highest median entropy as the root node. This is based on "a general trend that stem/progenitor populations tend to express a wider range of transcripts than do more differentiated cell types, along with higher entropy". But it is not a generalized trend, so if the researcher does not consider the root provided by algorithm is reasonable, he can re-specify a root/area to recalculate the pseudotime.  


```{r,echo=FALSE, results='asis'}
process <- function(x, k){
  marker <- x$dfgene$marker
  plot.gene <- split(marker$gene, marker$cluster)
  plot.gene <- lapply(plot.gene, function(i){i = i[1:k]})
  plot.gene <- c(do.call(rbind, plot.gene))
  plot.gene <- unique(plot.gene)
  x <- x$res
  x[[2]]$H.coords$factor <- paste0("F-", seq(1,12,1))
  x[[2]]$feature.coords <- subset(x[[2]]$feature.coords, name %in% plot.gene)
  x[[2]]$feature.coords$label <- paste0("G-", seq(1, length(plot.gene)))
  return(x)
}

SWNEplot <- function(cds, project){
  gsea.res <- cds[[2]]$gsea
  names(gsea.res) = paste0("F-", seq(1:12))
  gsea.filter <- lapply(gsea.res, function(x){x$pathway[1]})
  gsea.res <- lapply(gsea.res, function(x){
    x = subset(x, x$padj < 0.05)
  })
  openxlsx::write.xlsx(gsea.res, file = paste0(gsub("\\ ", "", project), ".xlsx"), overwrite = T)
  gsea.filter <- data.frame(
    row.names = NULL,
    "Factor" = paste0("F-", seq(1,12,1)),
    "Pathway" = do.call(rbind, gsea.filter)
  )
  gene <- cds[[2]]$feature.coords
  gene <- data.frame(
    "label" = gene$label,
    "gene" = gene$name
  )
  num <- dim(gene)[1]
  n <- ceiling(num/4)
  gene1 = gene[1:n,]; rownames(gene1) = NULL
  gene2 = gene[(n+1):(2*n),]; rownames(gene2) = NULL
  gene3 = gene[(2*n+1):(3*n),]; rownames(gene3) = NULL
  gene4 = gene[(3*n+1):num,]; rownames(gene4) = NULL

  plot1 <- plot_cells(cds[[1]], color_cells_by = "cell_type", label_cell_groups = F, label_branch_points = F, label_leaves = F, cell_size = 0.5) +
    geom_point(data = cds[[2]]$feature.coords, aes(x = x, y = y), 
             position = "identity", stat = "identity", shape = 17, size = 2) +
    geom_text_repel(fontface = "bold", data = cds[[2]]$feature.coords, aes(x = x, y = y, label = label))+
    ggtitle(paste0(project," trajectory on swne plot")) +
    theme(legend.position = "right", plot.title = element_text(hjust = 0.5))
  print(plot1)
  print(
    kable(list(gene1, matrix(numeric(), nrow = 0, ncol = 2), 
               gene2, matrix(numeric(), nrow = 0, ncol = 2), 
               gene3, matrix(numeric(), nrow = 0, ncol = 2), 
               gene4), align = "c") %>%
      kable_styling(full_width = T, position = "center")
  )

  plot2 <- plot_cells(cds[[1]], color_cells_by = "pseudotime", label_cell_groups = F, label_branch_points = F, label_leaves = F, cell_size = 0.5) +
    geom_point(data = cds[[2]]$H.coords, aes(x = x, y = y), 
               position = "identity", stat = "identity", shape = 17, size = 2) +
    geom_text(fontface = "bold", data = cds[[2]]$H.coords, aes(x = x, y = y + 0.03, label = factor))+
    ggtitle(paste0(project," pseudotime on swne plot")) +
    theme(legend.position = "right", plot.title = element_text(hjust = 0.5))
  print(plot2)

  return(list(gsea.filter = gsea.filter, gsea.res = gsea.res, gene = gene))
}
```

## {.tabset .tabset-pills}

### Fibroblast

```{r,echo=FALSE, results='asis'}
Fibro <- lapply(Fibro, function(x){process(x, 4)})
CT <- SWNEplot(Fibro$CT, paste0("Fibroblast (CT)"))
kable(CT$gsea.filter)
p53 <- SWNEplot(Fibro$p53, paste0("Fibroblast (p53)"))
kable(p53$gsea.filter)
```

### Myeloid

```{r,echo=FALSE, results='asis'}
Mye <- lapply(Mye, function(x){process(x, 4)})
CT <- SWNEplot(Mye$CT, paste0("Myeloid (CT)"))
kable(CT$gsea.filter)
p53 <- SWNEplot(Mye$p53, paste0("Myeloid (p53)"))
kable(p53$gsea.filter)
```

### Tumor

```{r,echo=FALSE, results='asis'}
Tumor <- lapply(Tumor, function(x){process(x, 4)})
CT <- SWNEplot(Tumor$CT, paste0("Tumor (CT)"))
kable(CT$gsea.filter)
p53 <- SWNEplot(Tumor$p53, paste0("Tumor (p53)"))
kable(p53$gsea.filter)
```